File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if(t == "") return "";

        unordered_map<char, int> countT;
        unordered_map<char, int> windowS;

        for(char chr : t){
            countT[chr]++;
        }

        int have = 0; 
        int need = countT.size();
        int l = 0;
        pair<int,int> response {-1, -1};
        int resp_length = INT_MAX;
        

        for(int r = 0; r < s.size(); r++){
            char c = s[r];
            windowS[c]++;

            if(countT.count(c) && windowS[c] == countT[c]){
                have++;
            }

            while(have == need){
                if((r - l + 1) < resp_length){
                    resp_length = r - l + 1;
                    response = {l, r};
                }

                windowS[s[l]]--;
                if(countT.count(s[l]) && windowS[s[l]] < countT[s[l]]){
                    have--;
                }
                l++;
            }
        }

        return resp_length == INT_MAX ? "" : s.substr(response.first, resp_length);
        
    }
};

File: Data Structures & Algorithms/permutation-string/submission-0.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        
        count1 = {}

        for let in s1: 
            count1[let] = 1 + count1.get(let, 0)

        need = len(count1)

        for i in range(len(s2)):
            count2 = {}
            count = 0
            for j in range(i, len(s2)): 
                count2[s2[j]] = 1 + count2.get(s2[j], 0)
                if count1.get(s2[j], 0) < count2[s2[j]]: 
                    break
                elif count1.get(s2[j], 0) == count2[s2[j]]: 
                    count+= 1
                if count == need: 
                    return True                
        return False





